### PR TITLE
Add Type Hints section to More Language Features lecture

### LINKED ISSUE #343: [New Lecture] Python Static Typing
It would be good to add a new lecture on `python` [type hints syntax](https://docs.python.org/3/library/typing.html). 

It has a number of good applications including:
 
1. improving efficiency of JIT compilers 
2. useful in software design projects to better record interface types for methods and functions
3. LLM's are now returning code with type hints, so teaching it will reduce confusion for new programmers
4. help to catch errors (in some IDE's) and using software like [pyrefly](https://pyrefly.org/)

Some notes when preparing the lecture

**numba:** to date Numba doesn't make use of them -- [instead you still need to use the Numba decorator](https://stackoverflow.com/questions/42310278/using-python-type-hints-with-numba) to declare types for JIT compilation.

**jax:** here is [the roadmap for python type hints](https://docs.jax.dev/en/latest/jep/12049-type-annotations.html) for the jax project

---

It could be either a new lecture, or as a new section in https://python-programming.quantecon.org/python_advanced_features.html

---

Things to check out:

1. [pyrefly](https://pyrefly.org/)
2. [warp kernel](https://nvidia.github.io/warp/modules/runtime.html#warp.kernel)
3. https://www.reddit.com/r/nvidia/comments/1jrfjkl/nvidia_adds_native_python_support_to_cuda/

### PR DESCRIPTION
**Pull request by mmcky**
## Summary

Adds a compact new section on Python type hints to the More Language Features lecture, placed between the Operators section and Decorators.

### Content

- Basic Syntax: annotating function parameters, return types, and variables
- Common Types: table of built-in types plus container type syntax
- Hints Do Not Enforce Types: demonstration that Python ignores type hints at runtime
- Why Use Type Hints: readability, IDE support, error checking tools, LLM-generated code
- Type Hints in Scientific Python: connection to JAX/Numba and JIT compilation concepts

### Design Decisions

- Kept minimal (~100 lines) to avoid overwhelming new Python programmers
- Added as a section (not a full lecture), per discussion in the issue
- Cross-references the existing dynamic typing discussion in need_for_speed
- All code cells execute successfully in the Jupyter Book build

Closes #343

### CODE REVIEW

**Comment by mmcky** [comment]
@jstac I think this is a nice (small) addition that captures the essence of your excellent comment in #343

**Comment by jstac** [comment]
Very nice :-)

BTW, I remember that last year the PRs were listing changed lectures (links to deployment lectures), making it easier to click through and see changes.  Is that infrastructure still available?

**Comment by mmcky** [comment]
nice catch @jstac -- it looks like I haven't updated lecture-python-programming. 

https://github.com/QuantEcon/meta/issues/296

Direct lecture linking is in python.quantecon.org repo (but not here). 

I'll roll out our quantecon/actions/netlify-preview action across the board. 

LGTM.

**Comment by mmcky** [comment]
- [x] do a pass of the quantecon style guide before merging. (@mmcky)

**Comment by jstac** [comment]
Thanks @mmcky 

Is there a way to automate this across the repos, along with updates to styles etc?  

Is it worth revisiting the one-big-repo idea?

Just tossing ideas around -- I don't really know the upsides and downsides...

**Comment by mmcky** [comment]
> Thanks @mmcky
> 
> Is there a way to automate this across the repos, along with updates to styles etc?
> 
> Is it worth revisiting the one-big-repo idea?
> 
> Just tossing ideas around -- I don't really know the upsides and downsides...

thanks @jstac 

I am actually running this PR over https://github.com/quantEcon/action-style-guide right now as a test case. 

This automation is still in work -- too much noise (I don't want to add noise -- only value).

**Comment by mmcky** [comment]
Applied **qe-writing-006** (sentence-case headings) across the entire lecture — not just the new Type Hints section. Fixed 14 headings total to capitalize only the first word and proper nouns.

**Comment by mmcky** [comment]
**Style fix (qe-writing-005):** Changed bold to italic in 4 places where formatting was used for emphasis rather than definitions:

- L25: *skip it on first pass* (pre-existing)
- Note box: *ignored by the Python interpreter at runtime*
- Type hints section: *not enforced*
- Type hints section: *clarity and tooling support*

Bold remains for the `**type hints**` definition and the numbered list labels (`**Readability**`, `**Editor support**`, etc.) which serve as definition-style labels.

**Comment by mmcky** [comment]
@jstac this looks good to me now. I'll merge this. thanks for casting an eye over it.

**Comment by jstac** [comment]
Thanks @mmcky, looks good.

File: lectures/python_advanced_features.md [hunk 1]
```diff
@@ -459,7 +459,106 @@ Overall, `*args` and `**kargs` are used when *defining a function*; they enable
...
+```{index} single: Python; Type Hints
+```
+
+Python is a *dynamically typed* language, meaning you don't need to declare the types of variables.
+
+(See our {doc}`earlier discussion <need_for_speed>` of dynamic versus static types.)
+
+However, Python supports optional **type hints** (also called type annotations) that allow you to indicate the expected types of variables, function parameters, and return values.
+
+Type hints were introduced in Python 3.5 and have become increasingly common in modern Python code.
```
[the remark this answers is not in the document]
Comment (mmcky) [reply lectures/python_advanced_features.md]: Good catch on the version nuance. We've updated the text to:

> Type hints were introduced starting in Python 3.5 and have evolved in subsequent versions. All of the syntax shown here works in Python 3.9 and later.

This keeps it accurate without overwhelming new Python programmers with the full PEP history (484/526/585). Since the course environment uses Python 3.13, all readers will have these features available.

File: lectures/python_advanced_features.md [hunk 2]
```diff
@@ -459,7 +459,106 @@ Overall, `*args` and `**kargs` are used when *defining a function*; they enable
...
+
+Python will not raise an error if you pass the "wrong" type:
+
+```{code-cell} python3
+def add(x: int, y: int) -> int:
+    return x + y
+
+# Works fine, despite passing strings
+add('foo', 'bar')
+```
```
[the remark this answers is not in the document]
Comment (mmcky) [reply lectures/python_advanced_features.md:539]: Agreed — the `add('foo', 'bar')` example was misleading because `+` is overloaded for strings, making it look like Python silently accepts any type mismatch.

We've replaced the example with floats-where-ints-are-hinted:

```python
def add(x: int, y: int) -> int:
    return x + y

# Passes floats — Python doesn't complain
add(1.5, 2.7)
```

Followed by: _The hints say `int`, but Python happily accepts `float` arguments and returns `4.2` --- also not an `int`._

This makes the mismatch unambiguous — both input and output type hints are visibly ignored, with no operator-overloading confusion.

[the remark this answers is not in the document]
Comment (mmcky) [reply lectures/python_advanced_features.md:539]: @jstac this was actually a great pickup from COPILOT review so I've changed the example.

File: lectures/python_advanced_features.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmcky) [reply lectures/python_advanced_features.md]: @jstac implemented this in a much simpler way (than the COPILOT feedback).

**Comment by mmcky** [comment]
\translate-resync

**Comment by mmcky** [comment]
### ✅ Translation sync completed (zh-cn)

**Target repo**: QuantEcon/lecture-python-programming.zh-cn
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.zh-cn/pull/2
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
### ✅ Translation sync completed (fa)

**Target repo**: QuantEcon/lecture-python-programming.fa
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.fa/pull/69
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
\translate-resync

**Comment by mmcky** [comment]
### ✅ Translation sync completed (fa)

**Target repo**: QuantEcon/lecture-python-programming.fa
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.fa/pull/70
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
### ✅ Translation sync completed (zh-cn)

**Target repo**: QuantEcon/lecture-python-programming.zh-cn
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.zh-cn/pull/3
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
\translate-resync

**Comment by mmcky** [comment]
### ✅ Translation sync completed (zh-cn)

**Target repo**: QuantEcon/lecture-python-programming.zh-cn
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.zh-cn/pull/6
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
### ✅ Translation sync completed (fa)

**Target repo**: QuantEcon/lecture-python-programming.fa
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.fa/pull/73
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
\translate-resync

**Comment by mmcky** [comment]
### ✅ Translation sync completed (zh-cn)

**Target repo**: QuantEcon/lecture-python-programming.zh-cn
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.zh-cn/pull/11
**Files synced** (1):
- `lectures/python_advanced_features.md`

**Comment by mmcky** [comment]
### ✅ Translation sync completed (fa)

**Target repo**: QuantEcon/lecture-python-programming.fa
**Translation PR**: https://github.com/QuantEcon/lecture-python-programming.fa/pull/78
**Files synced** (1):
- `lectures/python_advanced_features.md`
